src/common: add cephfs-shell section to cephfs-shell.conf

rishabh-d-dave opened the pull request:
This commit adds the ability to set cephfs-shell's config options
via ceph.conf as well as via ceph config subcommand.

Fixes: https://tracker.ceph.com/issues/44127



<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->
## Checklist
- [x] References tracker ticket
- [x] Updates documentation if necessary
- [x] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

liewegas commented:
```
/home/jenkins-build/build/workspace/ceph-pull-requests/src/test/cli/ceph-conf/invalid-args.t: failed
--- /home/jenkins-build/build/workspace/ceph-pull-requests/src/test/cli/ceph-conf/invalid-args.t
+++ /home/jenkins-build/build/workspace/ceph-pull-requests/src/test/cli/ceph-conf/invalid-args.t.err
@@ -8,7 +8,7 @@
   [1]
 
   $ ceph-conf -c test.conf --name total.garbage
-  error parsing 'total.garbage': expected string of the form TYPE.ID, valid types are: auth, mon, osd, mds, mgr, client
+  error parsing 'total.garbage': expected string of the form TYPE.ID, valid types are: auth, mon, osd, mds, mgr, client, cephfs-shell
   [1]
 ```

rishabh-d-dave commented:
I suppose this issue is not related to my PR since I could reproduce on latest master (079c1f9bfd4fb7aa2201960179235be1a4e3c2a9) too - 

```
$ ./bin/ceph-conf -c test.conf --name total.garbage
error parsing 'total.garbage': expected string of the form TYPE.ID, valid types are: auth, mon, osd, mds, mgr, client
```

tchaikov commented:
i cannot read the code block of "following options" in the ticket of https://tracker.ceph.com/issues/44127. @rishabh-d-dave could you update it to address this ?

rishabh-d-dave commented:
> i cannot read the code block of "following options" in the ticket of https://tracker.ceph.com/issues/44127. @rishabh-d-dave could you update it to address this ?

Added.

tchaikov commented:
i think this change is a follow-up of 33f5a59ac80262041b8dd19ebad6c10ba8d67e79 . 

personally, i'd avoid adding yet another entity type just for a certain client/tool. furthermore, i doubt if we should have these settings populated using `get_cephfs_shell_options()`. as, unlike other daemons / clients. cephfs-shell is not a distributed tool in the sense that its settings should be sync'ed across the cluster. instead, to my best understanding, it's like VIM or Bash, whose settings are consumed by the local user. actually, i'd suggest revert 33f5a59ac80262041b8dd19ebad6c10ba8d67e79  and instead of using `configparser` to read interested settings from `ceph.conf` or any specified conffile.

rishabh-d-dave commented:
> i think this change is a follow-up of [33f5a59](https://github.com/ceph/ceph/commit/33f5a59ac80262041b8dd19ebad6c10ba8d67e79) .
> 
> personally, i'd avoid adding yet another entity type just for a certain client/tool. furthermore, i doubt if we should have these settings populated using `get_cephfs_shell_options()`. as, unlike other daemons / clients. cephfs-shell is not a distributed tool in the sense that its settings should be sync'ed across the cluster. instead, to my best understanding, it's like VIM or Bash, whose settings are consumed by the local user. actually, i'd suggest revert [33f5a59](https://github.com/ceph/ceph/commit/33f5a59ac80262041b8dd19ebad6c10ba8d67e79) instead, and use `configparser` to read interested settings from `ceph.conf` or any specified conffile.

That was the original intent with PR #29964. See the first point here - https://github.com/ceph/ceph/pull/29964#issue-311892554. AFAIR, Patrick recommended to go this way instead, so that PR dropped using configparser altogether.

batrick commented:
> > i think this change is a follow-up of [33f5a59](https://github.com/ceph/ceph/commit/33f5a59ac80262041b8dd19ebad6c10ba8d67e79) .
> > personally, i'd avoid adding yet another entity type just for a certain client/tool. furthermore, i doubt if we should have these settings populated using `get_cephfs_shell_options()`. as, unlike other daemons / clients. cephfs-shell is not a distributed tool in the sense that its settings should be sync'ed across the cluster. instead, to my best understanding, it's like VIM or Bash, whose settings are consumed by the local user. actually, i'd suggest revert [33f5a59](https://github.com/ceph/ceph/commit/33f5a59ac80262041b8dd19ebad6c10ba8d67e79) instead, and use `configparser` to read interested settings from `ceph.conf` or any specified conffile.
> 
> That was the original intent with PR #29964. See the first point here - [#29964 (comment)](https://github.com/ceph/ceph/pull/29964#issue-311892554). AFAIR, Patrick recommended to go this way instead, so that PR dropped using configparser altogether.

I think the issue was that configparser was being used to parse `ceph.conf` which was redundant with the `ceph-common` parser.

I think I agree with Kefu's comment but I don't think we should keep putting cephfs-shell options in ceph.conf at all. Let's just have another conf file which by default is in `~/.cephfs-shell.conf`, or `$CEPHFS_SHELL_CONF`, or specified on the command line.

rishabh-d-dave commented:
> 
> I think I agree with Kefu's comment but I don't think we should keep putting cephfs-shell options in ceph.conf at all. Let's just have another conf file which by default is in `~/.cephfs-shell.conf`, or `$CEPHFS_SHELL_CONF`, or specified on the command line.

Wouldn't it better to keep all of Ceph's configuration option under one file? I think it would be easier to manage options that way. And just reading cephfs-shell section from ceph.conf would mostly be same as reading options from cephfs-shell.conf from the perspective of configparser.

tchaikov commented:
> > I think I agree with Kefu's comment but I don't think we should keep putting cephfs-shell options in ceph.conf at all. Let's just have another conf file which by default is in `~/.cephfs-shell.conf`, or `$CEPHFS_SHELL_CONF`, or specified on the command line.
> 
> Wouldn't it better to keep all of Ceph's configuration option under one file? I think it would be easier to manage options that way. And just reading cephfs-shell section from ceph.conf would mostly be same as reading options from cephfs-shell.conf from the perspective of configparser.

i think both options have their pros and cons:

- colocate cephfs specific option with ceph options: 
  * pros: one file to rule them all. and one parser to read all options. but please note, `configparser` is not equivalent to `ceph-common` parser, which is able to do meta variable substitution and more. we just need to make sure cephfs-shell does not rely on these features.
  * cons: user might be confused: why these options are not managed by `ceph config`?
- split them into two:
  * pros: more explicit. user will be able to edit the local conffile at will without duplicating the system wide `/etc/ceph/ceph.conf`. less confusion from user's perspective.
  * cons: yet another conf file. and two different ways to read options in source file.

i'd go with option 2.

rishabh-d-dave commented:
> pros: one file to rule them all. and one parser to read all options. but please note, configparser is not equivalent to ceph-common parser, which is able to do meta variable substitution and more. we just need to make sure cephfs-shell does not rely on these features.

We can parse ceph.conf using configparser instead of ceph-common parser.

liewegas commented:
I also prefer option 2

rishabh-d-dave commented:
Ok, I'll go for a separate conf file then.

tchaikov commented on src/tools/cephfs/cephfs-shell:
```diff
@@ -1504,36 +1507,44 @@ def get_bool_vals_for_boolopts(val):
+        if cp.has_option(sec, opt):
+            setattr(shell, opt, str_to_bool(cp.get(sec, opt)))
+
+def handle_conf(arg_conf=''):
```
nit, could use a better name than `handle_conf` , something like `get_shell_conffile_path()`.

rishabh-d-dave replied:
really. thanks! :)

tchaikov commented on src/tools/cephfs/cephfs-shell:
```diff
@@ -1556,10 +1569,10 @@ if __name__ == '__main__':
     shell = CephFSShell()
     shell.exit_code = 0
-    read_ceph_conf(shell, config_file)
+    read_shell_conf(shell, shell_conf_file)
```
```suggestion
    read_shell_conf(shell, handle_conf(args.config))
```

rishabh-d-dave replied:
yep, that's more compact. done.

tchaikov commented on src/tools/cephfs/cephfs-shell:
```diff
@@ -1556,10 +1569,10 @@ if __name__ == '__main__':
     sys.argv.extend([i.strip() for i in ' '.join(args.commands).split(',')])
 
-    setup_cephfs(config_file)
+    setup_cephfs(ceph_conf)
```
so `ceph_conf` is alway an empty string ?

rishabh-d-dave replied:
Incorporated what you suggested on IRC (got rid of the variable).

tchaikov commented on src/tools/cephfs/cephfs-shell:
```diff
@@ -1556,10 +1569,10 @@ if __name__ == '__main__':
-    setup_cephfs(config_file)
+    setup_cephfs(ceph_conf)
     shell = CephFSShell()
     shell.exit_code = 0
```
[off topic] why do we need to assign `0` to `shell.exit_code` here?

rishabh-d-dave replied:
there's reason behind but i can't recall it right now. i'll check again.

rishabh-d-dave commented on qa/tasks/cephfs/test_cephfs_shell.py:
```diff
@@ -28,49 +28,66 @@ def humansize(nbytes):
-    val = val.strip()
-    trueval = ['true', 'yes', 'y', '1']
-    return True if val == 1 or val.lower() in trueval else False
+    """
```
got rid of it.

rishabh-d-dave commented on qa/tasks/cephfs/test_cephfs_shell.py:
```diff
@@ -942,50 +961,63 @@ def test_help(self):
+    Contains tests for shell options from conf file and shell prompt.
+    """
+    def test_reading_conf(self):
+        editorval = self.get_cephfs_shell_cmd_output('set editor ?, set editor')
```
using `editor_val` instead.

rishabh-d-dave commented on qa/tasks/cephfs/test_cephfs_shell.py:
```diff
@@ -942,50 +961,63 @@ def test_help(self):
+        confcontents = "[cephfs-shell]\neditor = ???\neditor = %s\n" % (editorval)
+        sudo_write_file(self.mount_a.client_remote, tempconfpath, confcontents)
+
+        finaleditorval = self.get_cephfs_shell_cmd_output(
```
nice idea. added a sample output in comments - https://github.com/ceph/ceph/pull/33286/files#diff-0176720ea87b333b7c452e72f8adcf42R1004.

rishabh-d-dave commented on qa/tasks/cephfs/test_cephfs_shell.py:
```diff
@@ -942,50 +961,63 @@ def test_help(self):
+
+        tempconfpath = self.mount_a.client_remote.mktemp(
+            suffix='cephfs-shell.conf')
+        confcontents = "[cephfs-shell]\neditor = ???\neditor = %s\n" % (editorval)
```
i don't mind but i would like to keep it current way. it's more compact and cleaner since it's pretty short; above style violates indentation which i really don't like for strings as short as that.

varshar16 replied:
I would prefer readability with multiple line over style violation.

rishabh-d-dave commented on src/tools/cephfs/cephfs-shell:
```diff
@@ -1504,36 +1507,44 @@ def get_bool_vals_for_boolopts(val):
+        opts.append(attr)
+
+    # default and only section in our conf file.
+    cp = configparser.ConfigParser(default_section='cephfs-shell',
```
yep, done.

rishabh-d-dave commented on src/tools/cephfs/cephfs-shell:
```diff
@@ -1504,36 +1507,44 @@ def get_bool_vals_for_boolopts(val):
+        if cp.has_option(sec, opt):
+            setattr(shell, opt, str_to_bool(cp.get(sec, opt)))
+
+def handle_conf(arg_conf=''):
```
empty strings are fine but `None` is a problem for [`os.path.isfile()`](https://github.com/ceph/ceph/pull/33286/files#diff-9226892dd49173c4687a9953904ccaa8R1545)

rishabh-d-dave commented:
Tests were successful - http://pulpito.ceph.com/rishabh-2020-03-13_14:49:14-fs-wip-rishabh-fs-shell-set-opts-distro-basic-smithi/

gregsfortytwo commented:
One thing I notice here @rishabh-d-dave is that the Fixes: line is in the PR comment but not the commit text. It should go in the commit for easy tracking in git history, especially in case of backports.

rishabh-d-dave commented:
Thanks! Added "Fixes" line to the commit.

varshar16 commented on qa/tasks/cephfs/test_cephfs_shell.py:
```diff
@@ -82,7 +84,10 @@ def run_cephfs_shell_script(self, script, mount_x=None, stdin=None):
         mount_x.run_shell('chmod 755 ' + scriptpath)
 
-        args = ["cephfs-shell", "-c", mount_x.config_path, '-b', scriptpath]
+        args = ["cephfs-shell"]
+        if shell_conf_path:
+            args += ["-c", shell_conf_path]
+        args += ['-b', scriptpath]
```
```suggestion
        args = ["cephfs-shell", '-b', scriptpath]
        if shell_conf_path:
            args[1:1] = ["-c", shell_conf_path]
```

varshar16 commented on qa/tasks/cephfs/test_cephfs_shell.py:
```diff
@@ -943,50 +947,70 @@ def test_help(self):
+                                                      'editor')
+        editor_val = editor_val.replace("'", "", 2)
+        editor_val = editor_val.split(': ')[1]
+        new_editor_val = '???'
```
It can be a class attribute.

varshar16 commented on qa/tasks/cephfs/test_cephfs_shell.py:
```diff
@@ -943,50 +947,70 @@ def test_help(self):
+        self.assertEqual(editor_val, final_editor_val)
+
+    def test_setting_opt_after_reading_conf(self):
+        editor_val = self.get_cephfs_shell_cmd_output('set editor')
+        editor_val = editor_val.replace("'", "", 2)
+        editor_val = editor_val.split(': ')[1]
+        new_editor_val = '???'
+
+        tempconfpath = self.mount_a.client_remote.mktemp(
+            suffix='cephfs-shell.conf')
+        confcontents = "[cephfs-shell]\neditor = " + new_editor_val
+        sudo_write_file(self.mount_a.client_remote, tempconfpath,confcontents)
+        # output of the command executed in next line must look as follows -
+        # $ cephfs-shell -- set editor vim, set editor
+        # editor - was: vim
+        # now: vim
+        # editor: vim
+        final_editor_val = self.get_cephfs_shell_cmd_output(cmd=('set editor '
+            '%s, set editor' % (editor_val)), shell_conf_path=tempconfpath)
```
Also this code is repeated. Create a new method with cmd and confcontents as parameters?

rishabh-d-dave commented:
@gregsfortytwo Can you please do a final review? The teuthology tesitng was sucessful last time. See - http://pulpito.ceph.com/rishabh-2020-03-13_14:49:14-fs-wip-rishabh-fs-shell-set-opts-distro-basic-smithi/

gregsfortytwo commented:
There appear to be some unaddressed comments from @varshar16 and I don't think I'm likely to add anything in reviewing a code base written largely by the two of you. :)

rishabh-d-dave commented:
> There appear to be some unaddressed comments from @varshar16 and I don't think I'm likely to add anything in reviewing a code base written largely by the two of you. :)

Cool. I am putting this PR to test in that case.

varshar16 commented:
@rishabh-d-dave looks good. Can you please add a note in cephfs-shell doc about cephfs-shell.conf file?

tchaikov commented on src/tools/cephfs/cephfs-shell:
```diff
@@ -1504,36 +1507,52 @@ def get_bool_vals_for_boolopts(val):
+    cp.read(shell_conf_file)
+    for opt in opts:
+        if cp.has_option(sec, opt):
+            setattr(shell, opt, str_to_bool(cp.get(sec, opt)))
```
instead of using `str_to_bool()`, could use `cp.getboolean(sec, opt)` in a follow-up PR. see https://docs.python.org/3/library/configparser.html#configparser.ConfigParser.getboolean

rishabh-d-dave replied:
From the link you provided - `Any other value will cause it to raise ValueError. See get() for explanation of raw, vars and fallback`. This would lead to crash. str_to_bool returns bool if it can be converted to bool otherwise it returns string which is the required behaviour here.

rishabh-d-dave commented:
> Can you please add a note in cephfs-shell doc about cephfs-shell.conf file?

Done. PTAL.

varshar16 commented:
Looks fine. Can you add sample conf file in doc too?

rishabh-d-dave commented:
> Looks fine. Can you add sample conf file in doc too?

Skipping that since -
1. our conf follows standard INI format
2. we hardly have half a dozen of options
3. most of our options come from cmd2 which may change with it's version (which has happened pretty often)

varshar16 commented:
jenkins render docs

ceph-jenkins commented:
Doc render available at http://docs.ceph.com/ceph-prs/33286/

gregsfortytwo commented:
http://pulpito.front.sepia.ceph.com/gregf-2020-04-15_07:54:46-fs-wip-greg-testing-414-distro-basic-smithi/ went badly wrong for unrelated reasons, but it looks like http://pulpito.front.sepia.ceph.com/gregf-2020-04-15_07:54:46-fs-wip-greg-testing-414-distro-basic-smithi/4956501 is due to this PR @rishabh-d-dave?

rishabh-d-dave commented:
Jobs 4956575 and 4956501 must be due this PR. Other tests must not be affected by the code on this PR.

rishabh-d-dave commented:
http://pulpito.ceph.com/rishabh-2020-04-10_03:53:26-fs-wip-rishabh-fs-shell-set-opts-distro-basic-smithi/
http://pulpito.ceph.com/rishabh-2020-04-11_05:46:34-fs-wip-rishabh-fs-shell-set-opts-distro-basic-smithi/

rishabh-d-dave commented:
http://pulpito.ceph.com/rishabh-2020-04-18_04:14:21-fs-wip-rishabh-fs-shell-set-opts-distro-basic-smithi/

gregsfortytwo commented:
jenkins test make check

rishabh-d-dave commented:
Tested again this PR again after rebase; tests ran successfully - http://pulpito.ceph.com/rishabh-2020-04-28_05:08:49-fs-wip-rishabh-fs-shell-set-opts-distro-basic-smithi

rishabh-d-dave commented:
@gregsfortytwo I've updated the labels.

I'm not sure about adding needs-qa (I did that because you do it usually for every PR that has an approval). cephfs-shell is only tested with test_cephfs_shell.py so running integration tests again doesn't make sense IMHO. However, I don't mind waiting for another run. :)

varshar16 commented:
jenkins retest this please

varshar16 commented:
Running the make check again as the flake8 ci testing is fixed now.

varshar16 commented:
jenkins retest this please

varshar16 commented:
jenkins retest this please

varshar16 commented:
```
cephfs-shell:1517:1: E302 expected 2 blank lines, found 1
cephfs-shell:1548:1: E302 expected 2 blank lines, found 1
cephfs-shell:1564:1: E305 expected 2 blank lines after class or function definition, found 1
ERROR: InvocationError for command /home/jenkins-build/build/workspace/ceph-pull-requests/src/tools/cephfs/.tox/py3/bin/flake8 --ignore=W503 --max-line-length=100 cephfs-shell (exited with code 1)
```

rishabh-d-dave commented:
Couldn't reproduce those errors locally, but I've fixed those issues and pushed again.

rishabh-d-dave commented:
Tests were successful - http://pulpito.ceph.com/rishabh-2020-05-07_13:59:57-fs-wip-rishabh-fs-shell-set-opts-distro-basic-smithi
Link to the branch - https://github.com/rishabh-d-dave/ceph/commits/fs-shell-set-opts-test

gregsfortytwo commented:
@varshar16 if this is good to go can you put in a new review?

rishabh-d-dave commented:
@gregsfortytwo @batrick Can we please merge this PR?

rishabh-d-dave commented:
@batrick @gregsfortytwo ping.

batrick commented on src/tools/cephfs/cephfs-shell:
```diff
@@ -1512,37 +1515,61 @@ def get_bool_vals_for_boolopts(val):
+    env_conf = os.environ[env_var] if env_var in os.environ else ''
+
+    # here's the priority by which conf gets read.
+    for path in (arg_conf, env_conf, cwd_conf, home_dir_conf):
```
let's get rid of `cwd_conf`, that's unexpected for most unix applications.

rishabh-d-dave replied:
Ok. Fixed.

batrick commented on doc/cephfs/cephfs-shell.rst:
```diff
@@ -3,16 +3,22 @@
+Behaviour of CephFS Shell can be tweaked using cephfs-shell.conf. CephFS Shell
+looks for it by default at the path provided in environment variable
+`CEPHFS_SHELL_CONF`, then in current working directory and then in user's home
+directory.
```
give the name of that file in the home directory

rishabh-d-dave replied:
I am not sure what you mean but I have updated that line. Please take a look and let me know - https://github.com/ceph/ceph/pull/33286/files?file-filters%5B%5D=.py&file-filters%5B%5D=.rst&file-filters%5B%5D=No+extension#diff-f762f1d9ff1b1e7af6492d399e35b748R12

rishabh-d-dave commented:
> also need to remove the cephfs-shell options added to src/common/options.cc

Removed.
